File: openrest4js/openrest4js-api/Item.js
function ItemFromObj(data)
{
    return new Item(data.id, data.restaurantId, data.title, data.description, data.price, 
            data.variations ? ORJSArrayFromJsonArray(data.variations, VariationFromObj) : undefined, 
            data.availability ? AvailabilityFromObj(data.availability) : undefined, data.inactive, data.externalIds, 
            data.picture, data.labels);
}

function Item(id, restaurantId, title, description, price, variations, availability,
            inactive, externalIds, picture, labels)
{
    this.id = id;
    this.restaurantId = restaurantId;
    this.title = title;
    this.description = description;
    this.price = price ? price : 0;
    this.variations = variations ? variations : [];
    this.availability = availability ? availability : new Availability([], []);
    this.inactive = inactive ? inactive : false;
    this.externalIds = externalIds ? externalIds : {};
    this.picture = picture;
    this.labels = labels ? labels : [];
}

Item.prototype.toOpenRestObj = function()
{
    var ret = {};

    if (this.id) ret['id'] = this.id;
    if (this.restaurantId) ret['restaurantId'] = this.restaurantId;
    if (this.title) ret['title'] = this.title;
    if (this.description) ret['description'] = this.description;
    if (this.price) ret['price'] = this.price;
    if (this.variations) ret['variations'] = ORJsonArrayFromJSArray(this.variations);
    if (this.availability) ret['availability'] = this.availability.toOpenRestObj();
    if (this.inactive) ret['inactive'] = this.inactive;
    if (this.externalIds) ret['externalIds'] = this.externalIds;
    if (this.labels) ret['labels'] = this.labels;

    return ret;
}

File: openrest4cocoa/openrest4cocoa/openrest4cocoa-api/Item.m
//
//  Item.m
//  openrest4cocoa
//
//  Created by Yoav Amit on 5/4/11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "Item.h"
#import "Utils.h"

@implementation Item

@synthesize itemId;
@synthesize restaurantId;
@synthesize title;
@synthesize itemDescription;
@synthesize price;
@synthesize variations;
@synthesize availability;
@synthesize inactive;
@synthesize picture;

-(id)init
{
    if ((self = [super init]))
    {
        [self setVariations:[NSArray array]];
        [self setAvailability:[[[OpenrestAvailability alloc] init] autorelease]];
        [self setPrice:0];
        [self setInactive:false];
    }
    return self;
}

-(id)initWithDictionary:(NSDictionary*)data
{
    if ((self = [self init]))
    {
        [self setItemId:[data valueForKey:@"id"]];
        [self setRestaurantId:[data valueForKey:@"restaurantId"]];
        [self setTitle:[LocalizedDictionary dictionaryWithDictionary:[data valueForKey:@"title"]]];
        [self setItemDescription:[LocalizedDictionary dictionaryWithDictionary:[data valueForKey:@"description"]]];
        if ([data valueForKey:@"price"])
        {
            [self setPrice:[[data valueForKey:@"price"] intValue]];
        }
        if ([data valueForKey:@"variations"])
        {
            [self setVariations:[Utils refactorJsonArray:[data valueForKey:@"variations"]
                                                 toClass:@"Variation"]];
        }
        if ([data valueForKey:@"availability"])
        {
            [self setAvailability:[[[OpenrestAvailability alloc] initWithDictionary:[data valueForKey:@"availability"]] autorelease]];
        }
        if ([data valueForKey:@"inactive"])
        {
            [self setInactive:([[data valueForKey:@"inactive"] intValue] == 1)];
        }
        [self setPicture:[data valueForKey:@"picture"]];
    }
    return self;
}

-(NSUInteger)hash
{
    return [itemId hash];
}

-(BOOL)isEqual:(id)object
{
    if (object == self) return TRUE;
    if (object == NULL) return FALSE;
    if (![object isKindOfClass:[self class]]) return FALSE;

    Item* other = (Item*)object;
    if (![itemId isEqualToString:other.itemId]) return FALSE;
    if (![restaurantId isEqualToString:other.restaurantId]) return FALSE;
    if (![title isEqual:other.title]) return FALSE;
    if (![itemDescription isEqual:other.itemDescription]) return FALSE;
    if (price != other.price) return FALSE;
    if ((variations != other.variations) && (![variations isEqualToArray:other.variations])) return FALSE;
    if ((availability !=other.availability) && (![availability isEqual:other.availability])) return FALSE;
    if (inactive != other.inactive) return FALSE;
    if ((picture != other.picture) && (![picture isEqualToString:other.picture])) return FALSE;
    
    return TRUE;
}

-(NSDictionary*)proxyForJson
{
    NSMutableDictionary* ret = [NSMutableDictionary dictionaryWithCapacity:0];

    if (itemId != nil) {[ret setValue:itemId forKey:@"id"];}
    if (restaurantId != nil) {[ret setValue:restaurantId forKey:@"restaurantId"];}
    if (title != nil) {[ret setValue:title forKey:@"title"];}
    if (itemDescription != nil) {[ret setValue:itemDescription forKey:@"description"];}
    [ret setValue:[NSNumber numberWithInt:price] forKey:@"price"];
    if (variations != nil) {[ret setValue:variations forKey:@"variations"];}
    if (availability != nil) {[ret setValue:availability forKey:@"availability"];}
    [ret setValue:[NSNumber numberWithBool:inactive] forKey:@"inactive"];
    if (picture != nil) {[ret setValue:picture forKey:@"picture"];}

    
    return ret;
}

-(void)dealloc
{
    [itemId release];
    [restaurantId release];
    [title release];
    [itemDescription release];
    [variations release];
    [availability release];
    [picture release];
    [super dealloc];
}
@end
